packages/datadog-core/src/storage/async_resource.js

```diff
@@ -5,6 +5,7 @@ const { channel } = require('../../../diagnostics_channel')
 
 const beforeCh = channel('dd-trace:storage:before')
 const afterCh = channel('dd-trace:storage:after')
+const enterCh = channel('dd-trace:storage:enter')
```

Comment:
Just FYI, we're aiming long-term to eliminate use of async_hooks, so it would be ideal if we can eliminate these async_hooks lifecycle event channels. Not going to block this now as this will be a longer-term effort, but it's worth thinking about how you can do this without these channels.

Reply:
`dd-trace:storage:enter` is not async hook specific per se, its event fires when the value in the DD wrapper for `AsyncResourceStorage` is synchronously modified. I know `AsyncResourceStorage` is part of the async hooks library, but if we started using Node's `AsyncLocalStorage` under the hood for DD storage, we could still be emitting this enter event.

That said, I don't see us being able to avoid observing async context switches with the `before` event as we need to capture the active span on every context switch in case we need to associate it with a taken sample.

Reply:
Yep, it will be possible to wrap the `run` method for our use of `AsyncLocalStorage` to get the enter event, but the `before` event will likely go away or change substantially at some point so we should be thinking about how we can achieve that in some other way.

It's worth noting that my current work rewriting `AsyncLocalStorage` is to implement it on the native side which _could_ mean we expose a native interface to it which could enable you to just track your data in an `AsyncLocalStorage` instance and access that from the C++ side on your end whenever you need it. It _would_ mean needing to enter the isolate so if you're doing that in a signal it probably needs verification that it can function safely. If you can review https://github.com/nodejs/node/pull/48528 to see if that works for you, that would be a big help.